File: ios/LibrarySaveToCameraRoll.m
//
//  LibrarySaveToCameraRoll.m
//  MediaLibrary
//

#import "LibrarySaveToCameraRoll.h"
#import "MediaAssetManager.h"
#import <Photos/Photos.h>
#import <MobileCoreServices/MobileCoreServices.h>

@implementation LibrarySaveToCameraRoll

+ (PHAssetMediaType)assetTypeForUri:(NSString *)uri {
    NSString *fileExtension = [uri pathExtension];
    CFStringRef fileUTI = UTTypeCreatePreferredIdentifierForTag(
        kUTTagClassFilenameExtension,
        (__bridge CFStringRef)fileExtension,
        NULL
    );

    if (!fileUTI) return PHAssetMediaTypeUnknown;

    PHAssetMediaType type = PHAssetMediaTypeUnknown;
    if (UTTypeConformsTo(fileUTI, kUTTypeImage)) {
        type = PHAssetMediaTypeImage;
    } else if (UTTypeConformsTo(fileUTI, kUTTypeMovie)) {
        type = PHAssetMediaTypeVideo;
    } else if (UTTypeConformsTo(fileUTI, kUTTypeAudio)) {
        type = PHAssetMediaTypeAudio;
    }

    CFRelease(fileUTI);
    return type;
}

+ (NSURL *)normalizeAssetURLFromUri:(NSString *)uri {
    if ([uri hasPrefix:@"/"]) {
        return [NSURL URLWithString:[NSString stringWithFormat:@"file://%@", uri]];
    }
    return [NSURL URLWithString:uri];
}

+ (void)saveBlockWithAssetType:(PHAssetMediaType)assetType
                           url:(NSURL *)url
                    collection:(PHAssetCollection * _Nullable)collection
                    completion:(void (^)(NSString * _Nullable error, NSString * _Nullable identifier))completion {
    __block NSString *assetId = nil;

    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
        PHAssetChangeRequest *changeRequest;
        if (assetType == PHAssetMediaTypeVideo) {
            changeRequest = [PHAssetChangeRequest creationRequestForAssetFromVideoAtFileURL:url];
        } else {
            changeRequest = [PHAssetChangeRequest creationRequestForAssetFromImageAtFileURL:url];
        }

        PHObjectPlaceholder *assetPlaceholder = changeRequest.placeholderForCreatedAsset;
        assetId = assetPlaceholder.localIdentifier;

        if (collection) {
            PHFetchResult *photosAsset = [PHAsset fetchAssetsInAssetCollection:collection options:nil];
            PHAssetCollectionChangeRequest *albumChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:collection assets:photosAsset];
            [albumChangeRequest addAssets:@[assetPlaceholder]];
        }
    } completionHandler:^(BOOL success, NSError *error) {
        if (error) {
            completion(error.localizedDescription, nil);
        } else {
            completion(nil, assetId);
        }
    }];
}

+ (void)saveWithOptionsAlbum:(NSString * _Nullable)album
                   assetType:(PHAssetMediaType)assetType
                         url:(NSURL *)url
                  completion:(void (^)(NSString * _Nullable error, NSString * _Nullable identifier))completion {
    if (!album || album.length == 0) {
        [self saveBlockWithAssetType:assetType url:url collection:nil completion:completion];
        return;
    }

    PHFetchOptions *fetchOptions = [[PHFetchOptions alloc] init];
    fetchOptions.predicate = [NSPredicate predicateWithFormat:@"title = %@", album];

    PHFetchResult *collections = [PHAssetCollection fetchAssetCollectionsWithType:PHAssetCollectionTypeAlbum
                                                                          subtype:PHAssetCollectionSubtypeAny
                                                                          options:fetchOptions];

    if (collections.firstObject) {
        [self saveBlockWithAssetType:assetType url:url collection:collections.firstObject completion:completion];
        return;
    }

    // Create new collection
    __block NSString *placeholderId = nil;

    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
        PHAssetCollectionChangeRequest *createAlbum = [PHAssetCollectionChangeRequest creationRequestForAssetCollectionWithTitle:album];
        placeholderId = createAlbum.placeholderForCreatedAssetCollection.localIdentifier;
    } completionHandler:^(BOOL success, NSError *error) {
        if (error || !placeholderId) {
            completion(error.localizedDescription ?: @"Failed to create album", nil);
            return;
        }

        PHFetchResult *newCollections = [PHAssetCollection fetchAssetCollectionsWithLocalIdentifiers:@[placeholderId] options:nil];
        PHAssetCollection *newCollection = newCollections.firstObject;

        if (newCollection) {
            [self saveBlockWithAssetType:assetType url:url collection:newCollection completion:completion];
        } else {
            completion(@"Failed to fetch created album", nil);
        }
    }];
}

+ (void)saveToCameraRollWithLocalUri:(NSString *)localUri
                               album:(NSString *)album
                            callback:(void (^)(NSString * _Nullable error, NSString * _Nullable result))callback {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if (![[NSBundle mainBundle] objectForInfoDictionaryKey:@"NSPhotoLibraryAddUsageDescription"]) {
            callback(@"kErrorNoPermissions", nil);
            return;
        }

        if ([localUri pathExtension].length == 0) {
            callback(@"kErrorNoFileExtension", nil);
            return;
        }

        PHAssetMediaType assetType = [self assetTypeForUri:localUri];
        if (assetType == PHAssetMediaTypeAudio || assetType == PHAssetMediaTypeUnknown) {
            callback(@"kErrorUnsupportedAsset", nil);
            return;
        }

        NSURL *assetUrl = [self normalizeAssetURLFromUri:localUri];
        if (!assetUrl) {
            callback(@"kErrorNotValidUri", nil);
            return;
        }

        [self saveWithOptionsAlbum:album assetType:assetType url:assetUrl completion:^(NSString *error, NSString *identifier) {
            if (error) {
                callback(error, nil);
                return;
            }

            if (identifier) {
                [MediaAssetManager fetchAssetWithIdentifier:identifier completion:^(NSString *data) {
                    if (data) {
                        callback(nil, data);
                    } else {
                        callback(@"kErrorFetchAssetData", nil);
                    }
                }];
                return;
            }

            callback(@"kErrorSaveToCameraRoll", nil);
        }];
    });
}

@end

File: ios/LibrarySaveToCameraRoll.swift
//
//  LibrarySaveToCameraRoll.swift
//  MediaLibrary
//
//  Created by sergeymild on 03/08/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import CoreServices
import Photos

class Result {
    internal init(error: String? = nil, result: String? = nil) {
        self.error = error
        self.result = result
    }
    
    var error: String?
    var result: String?
}

@objc
public class LibrarySaveToCameraRoll: NSObject {
    private static func assetTypeFor(uri: NSString) -> PHAssetMediaType {
        let fileExtension = uri.pathExtension
        guard let fileUTI = UTTypeCreatePreferredIdentifierForTag(
            kUTTagClassFilenameExtension,
            fileExtension as CFString,
            nil
        ) else { return .unknown }
        let value = fileUTI.takeRetainedValue()
        if (UTTypeConformsTo(value, kUTTypeImage)) { return .image }
        if (UTTypeConformsTo(value, kUTTypeMovie)) { return .video }
        if (UTTypeConformsTo(value, kUTTypeAudio)) { return .audio }
        return .unknown
    }
    
    private static func normalizeAssetURLFromUri(uri: NSString) -> URL? {
        if uri.hasPrefix("/") {
            return URL(string: "file://\(uri)")
        }
        return URL(string: uri as String)
    }
    
    private static func saveBlock(
        assetType: PHAssetMediaType,
        url: URL,
        collection: PHAssetCollection?
    ) async -> Result {
        let result = Result()
        do {
            try await PHPhotoLibrary.shared().performChanges {
            let changeRequest = assetType == .video
            ? PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: url)
            : PHAssetChangeRequest.creationRequestForAssetFromImage(atFileURL: url)
            
            let assetPlaceholder = changeRequest!.placeholderForCreatedAsset
            
            if let collection = collection {
                let photosAsset = PHAsset.fetchAssets(
                    in: collection,
                    options: nil
                )
                let albumChangeRequest = PHAssetCollectionChangeRequest(
                    for: collection,
                    assets: photosAsset
                )
                albumChangeRequest?.addAssets([assetPlaceholder] as NSFastEnumeration)
            }
            result.error = nil
            result.result = assetPlaceholder!.localIdentifier
        }
            
        }
        catch let error {
            print(error.localizedDescription)
            result.error = error.localizedDescription
        }
        return result
    }
    
    private static func saveWithOptions(
        album: String?,
        assetType: PHAssetMediaType,
        url: URL,
        collection: PHAssetCollection?
    ) async -> Result {
        if album == nil || album!.isEmpty {
            return await saveBlock(assetType: assetType, url: url, collection: collection)
        }
        
        let fetchOptions = PHFetchOptions()
        fetchOptions.predicate = NSPredicate(format: "title = %@", argumentArray: [album!])
        
        let collection = PHAssetCollection.fetchAssetCollections(
            with: .album,
            subtype: .any,
            options: fetchOptions
        )
        if let collection = collection.firstObject {
            return await saveBlock(assetType: assetType, url: url, collection: collection)
        }
        
        // create new collection
        var placeholder: PHObjectPlaceholder?
        
        let result = Result()
        do {
            try await PHPhotoLibrary.shared().performChanges {
                let createAlbum = PHAssetCollectionChangeRequest.creationRequestForAssetCollection(
                    withTitle: album!
                )
                placeholder = createAlbum.placeholderForCreatedAssetCollection
            }
            // if collection was created fetch it
            if let placeholder = placeholder {
                let collection = PHAssetCollection.fetchAssetCollections(
                    withLocalIdentifiers: [placeholder.localIdentifier],
                    options: nil
                ).firstObject
                if let collection = collection {
                    return await Self.saveBlock(
                        assetType: assetType,
                        url: url,
                        collection: collection
                    )
                }
            }
        }
        catch let error {
            print(error.localizedDescription)
            result.error = error.localizedDescription
        }
        
        return result
    }
    
    @objc
    public static func saveToCameraRoll(
        localUri: NSString,
        album: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        Task {
            if Bundle.main.object(forInfoDictionaryKey: "NSPhotoLibraryAddUsageDescription") == nil {
                return callback("kErrorNoPermissions", nil)
            }
            if localUri.pathExtension.isEmpty {
                return callback("kErrorNoFileExtension", nil);
            }
            
            let assetType = Self.assetTypeFor(uri: localUri)
            if assetType == .audio || assetType == .unknown {
                return callback("kErrorUnsupportedAsset", nil)
            }
            
            guard let assetUrl = Self.normalizeAssetURLFromUri(uri: localUri) else {
                return callback("kErrorNotValidUri", nil)
            }
            
            let result = await saveWithOptions(
                album: album,
                assetType: assetType,
                url: assetUrl,
                collection: nil
            )
            if let error = result.error {
                return callback(error, nil)
            }
            if let id = result.result {
                return MediaAssetManager.fetchAsset(identifier: id) { data in
                    if let data = data {
                        return callback(nil, data)
                    }
                    callback("kErrorFetchAssetData", nil)
                }
            }
            return callback("kErrorSaveToCameraRoll", nil)
        }
    }
}
